tensorflow_probability/python/experimental/bijectors/highway_flow.py

```diff
@@ -0,0 +1,376 @@
+        x = tf.reshape(x, -1)
+      added_batch = True
+      x = tf.expand_dims(x, 0)
+    fldj = tf.zeros(x.shape[:-1]) + tf.reduce_sum(
+      tf.math.log(tf.concat([(self.residual_fraction) * tf.ones(
+        self.gate_first_n), tf.zeros(self.width - self.gate_first_n)],
+                            axis=0) + (
+                    tf.concat([(1. - self.residual_fraction) * tf.ones(
+                      self.gate_first_n),
+                               tf.ones(self.width - self.gate_first_n)],
+                              axis=0)) * tf.linalg.diag_part(
+        self.upper_diagonal_weights_matrix)))
+    x = tf.linalg.matvec(
+      self._convex_update(self.lower_diagonal_weights_matrix), x)
+    x = tf.linalg.matvec(tf.transpose(
+      self._convex_update(self.upper_diagonal_weights_matrix)),
+      x)
+    x += tf.concat([(1. - self.residual_fraction) * tf.ones(
+      self.gate_first_n), tf.ones(self.width - self.gate_first_n)],
+                   axis=0) * self.bias
+
+    if self.activation_fn:
+      fldj += tf.reduce_sum(tf.math.log(self._derivative_of_softplus(x)),
+                            -1)
+      x = tf.concat([(self.residual_fraction) * tf.ones(
+        self.gate_first_n), tf.zeros(self.width - self.gate_first_n)],
+                    axis=0) * x + tf.concat(
+        [(1. - self.residual_fraction) * tf.ones(
+          self.gate_first_n), tf.ones(self.width - self.gate_first_n)],
+        axis=0) * tf.nn.softplus(x)
+    if added_batch:
+      x = tf.squeeze(x, 0)
+    return x, {'ildj': -fldj, 'fldj': fldj}
+
+  def _augmented_inverse(self, y):
+    """Computes inverse and inverse_log_det_jacobian transformations.
+
+    Args:
+      y: input of the (inverse) bijectorr.
+
+    Returns:
+      y after inverse flow and a dict containing inverse and forward log
+      determinant of the jacobian.
+    """
+
+    added_batch = False
+    if len(y.shape) <= 1:
+      if len(y.shape) == 0:
+        y = tf.reshape(y, -1)
+      added_batch = True
+      y = tf.expand_dims(y, 0)
+    ildj = tf.zeros(y.shape[:-1]) - tf.reduce_sum(
+      tf.math.log(tf.concat([(self.residual_fraction) * tf.ones(
+        self.gate_first_n), tf.zeros(self.width - self.gate_first_n)],
+                            axis=0) + tf.concat(
+        [(1. - self.residual_fraction) * tf.ones(
+          self.gate_first_n), tf.ones(self.width - self.gate_first_n)],
+        axis=0) * tf.linalg.diag_part(
+        self.upper_diagonal_weights_matrix)))
+    if self.activation_fn:
+      y = self._inverse_of_softplus(y)
+      ildj -= tf.reduce_sum(tf.math.log(self._derivative_of_softplus(y)),
+                            -1)
+
+    y = y - tf.concat([(1. - self.residual_fraction) * tf.ones(
+      self.gate_first_n), tf.ones(self.width - self.gate_first_n)],
+                      axis=0) * self.bias
+    y = tf.linalg.triangular_solve(tf.transpose(
+      self._convex_update(self.upper_diagonal_weights_matrix)),
+      tf.linalg.matrix_transpose(y),
```

Comment:
Why are we transposing `y` here? Transposes are expensive, since they actually change the data layout in memory---as opposed to other ops like reshape, expand_dims, squeeze, etc., which just change metadata---so it's good to avoid them when possible. (also it just feels weird to be transposing `y` since it's conceptually a vector).

I think you can remove all `transpose` and `matrix_transpose` calls from this method, using a combination of:

1) Work with `y` of shape `[..., width, 1]` throughout, so that you can feed it directly to `triangular_solve` calls. That is, instead of `y = expand_dims(y, 0)` at the beginning, do `y = y[..., tf.newaxis]` to add the extra dimension at the end rather than at the beginning. (as a bonus, you can do this unconditionally so you don't need the `added_batch` logic any more). You might need to sprinkle in a few more `[..., tf.newaxis]` expansions for other Tensors, eg `bias`, to get the dimensions to line up, but this is costless.

2. Use the `adjoint` flag to `triangular_solve`. Since transposes are expensive, most TF matrix ops give you the option to treat a matrix as transposed without explicitly constructing the transpose. For example, `triangular_solve(A, B, adjoint=True)` is equivalent to, and more efficient than, `triangular_solve(tf.linalg.matrix_transpose(A), B)`.